From: Tanu Kaskinen <tanu.kaskinen@linux.intel.com>
Date: Mon, 4 Aug 2014 21:26:17 +0300
Subject: sink-input, source-output: Add hooks for reference ratio changes

Needed for implementing relative volume controls for streams in
module-volume-api. The plan is to create those controls in the core,
though, and these hooks won't be needed at that point any more.

Change-Id: Id30f38f4adfa9ede7bd0b12b484fe329ca1a3991
---
 src/pulsecore/core.h          | 2 ++
 src/pulsecore/sink-input.c    | 2 ++
 src/pulsecore/sink-input.h    | 3 ++-
 src/pulsecore/source-output.c | 2 ++
 src/pulsecore/source-output.h | 3 ++-
 5 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/pulsecore/core.h b/src/pulsecore/core.h
index 1f042b9..0e8f709 100644
--- a/src/pulsecore/core.h
+++ b/src/pulsecore/core.h
@@ -100,6 +100,7 @@ typedef enum pa_core_hook {
     PA_CORE_HOOK_SINK_INPUT_STATE_CHANGED,
     PA_CORE_HOOK_SINK_INPUT_PROPLIST_CHANGED,
     PA_CORE_HOOK_SINK_INPUT_VOLUME_CHANGED,
+    PA_CORE_HOOK_SINK_INPUT_REFERENCE_RATIO_CHANGED,
     PA_CORE_HOOK_SINK_INPUT_MUTE_CHANGED,
     PA_CORE_HOOK_SINK_INPUT_SEND_EVENT,
     PA_CORE_HOOK_SOURCE_OUTPUT_NEW,
@@ -113,6 +114,7 @@ typedef enum pa_core_hook {
     PA_CORE_HOOK_SOURCE_OUTPUT_STATE_CHANGED,
     PA_CORE_HOOK_SOURCE_OUTPUT_PROPLIST_CHANGED,
     PA_CORE_HOOK_SOURCE_OUTPUT_VOLUME_CHANGED,
+    PA_CORE_HOOK_SOURCE_OUTPUT_REFERENCE_RATIO_CHANGED,
     PA_CORE_HOOK_SOURCE_OUTPUT_MUTE_CHANGED,
     PA_CORE_HOOK_SOURCE_OUTPUT_SEND_EVENT,
     PA_CORE_HOOK_CLIENT_NEW,
diff --git a/src/pulsecore/sink-input.c b/src/pulsecore/sink-input.c
index 57c906d..d62be6f 100644
--- a/src/pulsecore/sink-input.c
+++ b/src/pulsecore/sink-input.c
@@ -2352,4 +2352,6 @@ void pa_sink_input_set_reference_ratio(pa_sink_input *i, const pa_cvolume *ratio
     pa_log_debug("Sink input %u reference ratio changed from %s to %s.", i->index,
                  pa_cvolume_snprint_verbose(old_ratio_str, sizeof(old_ratio_str), &old_ratio, &i->channel_map, true),
                  pa_cvolume_snprint_verbose(new_ratio_str, sizeof(new_ratio_str), ratio, &i->channel_map, true));
+
+    pa_hook_fire(&i->core->hooks[PA_CORE_HOOK_SINK_INPUT_REFERENCE_RATIO_CHANGED], i);
 }
diff --git a/src/pulsecore/sink-input.h b/src/pulsecore/sink-input.h
index e5b0ae8..c99ce1f 100644
--- a/src/pulsecore/sink-input.h
+++ b/src/pulsecore/sink-input.h
@@ -440,7 +440,8 @@ void pa_sink_input_set_volume_direct(pa_sink_input *i, const pa_cvolume *volume)
 /* Called from the main thread, from sink.c only. This shouldn't be a public
  * function, but the flat volume logic in sink.c currently needs a way to
  * directly set the sink input reference ratio. This function simply sets
- * i->reference_ratio and logs a message if the value changes. */
+ * i->reference_ratio and logs a message and fires a hook if the value
+ * changes. */
 void pa_sink_input_set_reference_ratio(pa_sink_input *i, const pa_cvolume *ratio);
 
 #define pa_sink_input_assert_io_context(s) \
diff --git a/src/pulsecore/source-output.c b/src/pulsecore/source-output.c
index 0012be3..ae5a92c 100644
--- a/src/pulsecore/source-output.c
+++ b/src/pulsecore/source-output.c
@@ -1718,4 +1718,6 @@ void pa_source_output_set_reference_ratio(pa_source_output *o, const pa_cvolume
     pa_log_debug("Source output %u reference ratio changed from %s to %s.", o->index,
                  pa_cvolume_snprint_verbose(old_ratio_str, sizeof(old_ratio_str), &old_ratio, &o->channel_map, true),
                  pa_cvolume_snprint_verbose(new_ratio_str, sizeof(new_ratio_str), ratio, &o->channel_map, true));
+
+    pa_hook_fire(&o->core->hooks[PA_CORE_HOOK_SOURCE_OUTPUT_REFERENCE_RATIO_CHANGED], o);
 }
diff --git a/src/pulsecore/source-output.h b/src/pulsecore/source-output.h
index 3ed950b..60bbda8 100644
--- a/src/pulsecore/source-output.h
+++ b/src/pulsecore/source-output.h
@@ -362,7 +362,8 @@ void pa_source_output_set_volume_direct(pa_source_output *o, const pa_cvolume *v
 /* Called from the main thread, from source.c only. This shouldn't be a public
  * function, but the flat volume logic in source.c currently needs a way to
  * directly set the source output reference ratio. This function simply sets
- * o->reference_ratio and logs a message if the value changes. */
+ * o->reference_ratio and logs a message and fires a hook if the value
+ * changes. */
 void pa_source_output_set_reference_ratio(pa_source_output *o, const pa_cvolume *ratio);
 
 #define pa_source_output_assert_io_context(s) \
